crates/bounded_planet_core/src/networking/systems.rs

```diff
@@ -117,190 +123,186 @@ pub fn receive_net_events(
-    tokio::spawn(send_to_streams(cid, connection, recv, event_sender));
-}
-
-async fn handle_incoming_streams(
-    mut uni_streams: IncomingUniStreams,
-    event_sender: UnboundedSender<ReceiveEvent>,
-    id: ConnectionId
-) {
-    // Keep getting events from the connection until it closes
-    while let Some(stream) = uni_streams.next().await {
-        match stream {
-            Err(quinn::ConnectionError::ApplicationClosed { .. }) => break,
+        // Wait for connection to finish connecting
+        let quinn::NewConnection { connection, uni_streams, .. } = match self.connecting.await {
+            Ok(connection) => connection,
             Err(e) => {
-                event_sender.send(ReceiveEvent::NetworkError(
+                self.event_sender.send(ReceiveEvent::NetworkError(
                     NetworkError::ConnectionError(e)
                 )).expect("Failed to send network event");
-                break;
-            },
-            Ok(recv) => {
-                tokio::spawn(read_from_stream(id, recv, event_sender.clone()));
+                return;
             }
         };
-    }
 
-    // Send a final event indicating that this connection closed
-    event_sender
-        .send(ReceiveEvent::Disconnected(id))
-        .expect("Failed to send network event");
+        // Create a new MPSC which the ECS can use to send packets through this connection
+        let (send, recv) = unbounded_channel();
+
+        // Send an intial event indicating that this connection opened
+        self.event_sender
+            .send(ReceiveEvent::Connected(self.id, send))
+            .expect("Failed to send network event");
+
+        // Start running tasks to send/receive to this connection
+        tokio::spawn(Connected::new(
+            self.id,
+            connection,
+            uni_streams,
+            recv,
+            self.event_sender,
+        ).run());
+    }    
 }
 
-/// Handle all the work of a specific stream
-async fn read_from_stream(
-    connection_id: ConnectionId,
-    stream_recv: RecvStream<TlsSession>,
-    event_sender: UnboundedSender<ReceiveEvent>,
-) {
-    info!("Stream incoming: conn:{:?}", connection_id);
-
-    // Pull packets from this stream and publish them to the ECS through the event_sender
-    let mut recv = BoundedPlanetRecvStream::new(stream_recv);
-    loop {
-        let pkt = recv.recv_packet().await;
-        let se = match pkt {
-            Ok(pkt) => event_sender.send(ReceiveEvent::ReceivedPacket {
-                connection: connection_id,
-                data: Arc::new(pkt),
-            }),
-            Err(err) => {
-                event_sender.send(ReceiveEvent::NetworkError(
-                    NetworkError::ReceiveError {
-                        connection: connection_id,
-                        err,
-                    }
-                )).expect("Failed to send error event!");
+/// Represents an active quinn connection
+struct Connected {
```

Comment:
Based on the comment and usage, this is really our own logical wrapper around a running connection. Based on this, the name `Connected` is somewhat confusing. Initially I was confused based on the name, motivating this.

Reply:
This is one of the most actually important suggestions I feel.